### PR TITLE
[WIP] add directive for frankenphp

### PR DESCRIPTION
**Pull request by flug**

### CODE REVIEW

File: caddy/caddy.go [hunk 1]
```diff
@@ -243,6 +249,186 @@ func parseCaddyfile(h httpcaddyfile.Helper) (caddyhttp.MiddlewareHandler, error)
...
+				args := dispenser.RemainingArgs()
+				dispenser.Delete()
+				for range args {
+					dispenser.Delete()
+				}
+				if len(args) != 2 {
+					return nil, dispenser.ArgErr()
+				}
+				if frankenphpTransport.Env == nil {
+					frankenphpTransport.Env = make(map[string]string)
```
Comment (dunglas) [inline caddy/caddy.go:296]: You could probably initialize the map outside of the loop to simplify the code and remove this `if`.

File: caddy/caddy.go [hunk 2]
```diff
@@ -366,41 +365,15 @@ func parseFrankenPHP(h httpcaddyfile.Helper) ([]httpcaddyfile.ConfigValue, error
...
-	// create the final reverse proxy route which is
-	// conditional on matching PHP files
-	rpRoute := caddyhttp.Route{
-		MatcherSetsRaw: []caddy.ModuleMap{rpMatcherSet},
-		HandlersRaw:    []json.RawMessage{caddyconfig.JSONModuleObject(rpHandler, "handler", "reverse_proxy", nil)},
-	}
 
 	subroute := caddyhttp.Subroute{
-		Routes: append(routes, rpRoute),
+		Routes: append(routes),
```
Comment (dunglas) [inline caddy/caddy.go:376]:
```suggestion
		Routes: routes,
```

?

File: caddy/caddy.go [hunk 3]
```diff
@@ -243,6 +248,159 @@ func parseCaddyfile(h httpcaddyfile.Helper) (caddyhttp.MiddlewareHandler, error)
 	return m, err
 }
 
+func parseFrankenPHP(h httpcaddyfile.Helper) ([]httpcaddyfile.ConfigValue, error) {
+
+	if !h.Next() {
+		return nil, h.ArgErr()
+	}
+
+	frankenphpTransport := FrankenPHPModule{}
```
Comment (dunglas) [inline caddy/caddy.go:257]:
```suggestion
	frankenphp := FrankenPHPModule{}
```

File: testdata/Caddyfile [hunk 4]
```diff
@@ -23,6 +23,11 @@ http:// {
 		}
 		rewrite @indexFiles {http.matchers.file.relative}
 
+	@frankenphp path *.php {
+		root public/
+		index {path}/index.php 
+		resolve_root_symlink
+	}
```
Comment (dunglas) [inline testdata/Caddyfile:30]: So the following lines can be removed, right?

Also, the indentation doesn't look good (you should be able to fix this automatically with [`caddy fmt`](https://caddyserver.com/docs/command-line#caddy-fmt).

**Comment by flug** [comment]
last error : ```Error: adapting config using caddyfile: directive 'frankenphp' is not an ordered HTTP handler, so it cannot be used here```

File: caddy/caddy.go [hunk 5]
```diff
@@ -151,7 +152,10 @@ type FrankenPHPModule struct {
 	SplitPath          []string          `json:"split_path,omitempty"`
```
Comment (dunglas) [inline caddy/caddy.go:152]: Shouldn't it be `Split` for consistency with the `php_fastcgi` directive?

File: caddy/frankenphp/Caddyfile [hunk 6]
```diff
@@ -1,37 +1,18 @@
 {
 	# Debug
-    {$CADDY_DEBUG}
-
-	frankenphp {
-		#worker /path/to/your/worker.php
+	{$CADDY_DEBUG}
+    frankenphp index {
```
Comment (dunglas) [inline caddy/frankenphp/Caddyfile:4]: indentation to fix.

Shouldn't be this directive outside of the global block (as `php_fastcgi`)?

Also, why do we need to pass the `index` matcher? It doesn't look defined.

Also, by default just having `frankenphp` should be enough as for `php_fastcgi`.

File: testdata/Caddyfile [hunk 7]
```diff
@@ -1,34 +1,10 @@
...
-		file_server
-
-		respond 404
-	}
+    debug
+    frankenphp index {
+        root /path/to/root
+        split_path /index.php
+        resolve_root_symlink true
+
```
Comment (dunglas) [inline testdata/Caddyfile:7]:
```suggestion
```

**Review by dunglas — CHANGES_REQUESTED** [review]

**Pull request closed without merging**